op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -0,0 +1,440 @@
+		db.CreateAccount(k)
+
+		db.SetNonce(k, v.Nonce)
+		db.SetBalance(k, v.Balance)
+		db.SetCode(k, v.Code)
+		db.SetStorage(k, v.Storage)
+
+		log.Info("Moved account", "address", k)
+		balanceDiff = balanceDiff.Add(balanceDiff, v.Balance)
+	}
+	log.Info("Migrated OP contracts into state DB", "copiedAccounts", accountCounter, "overwrittenAccounts", overwriteCounter)
+	if balanceDiff.Sign() != 0 {
+		log.Warn("Deploying OP contracts changed native balance", "diff", balanceDiff)
+	}
+
+	// We're done messing around with the database, so we can now commit the changes to the DB.
+	// Note that this doesn't actually write the changes to disk.
+	log.Info("Committing state DB")
+	// TODO(pl): What block info to put here?
+	newRoot, err := db.Commit(1234, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the header for the Bedrock transition block.
+	cel2Header := &types.Header{
+		ParentHash:  header.Hash(),
+		UncleHash:   types.EmptyUncleHash,
+		Coinbase:    predeploys.SequencerFeeVaultAddr,
+		Root:        newRoot,
+		TxHash:      types.EmptyRootHash,
+		ReceiptHash: types.EmptyRootHash,
+		Bloom:       types.Bloom{},
+		Difficulty:  new(big.Int).Set(common.Big0),
+		Number:      new(big.Int).Add(header.Number, common.Big1),
+		GasLimit:    header.GasLimit,
+		GasUsed:     0,
+		Time:        uint64(time.Now().Unix()), // TODO(pl): Needed to avoid L1-L2 time mismatches
+		Extra:       []byte("CeL2 migration"),
+		MixDigest:   common.Hash{},
+		Nonce:       types.BlockNonce{},
+		BaseFee:     new(big.Int).Set(header.BaseFee),
+	}
+	log.Info("Build Cel2 migration header", "header", cel2Header)
+
+	// Create the Bedrock transition block from the header. Note that there are no transactions,
+	// uncle blocks, or receipts in the Bedrock transition block.
+	cel2Block := types.NewBlock(cel2Header, nil, nil, nil, trie.NewStackTrie(nil))
+
+	// We did it!
+	log.Info(
+		"Built Cel2 migration block",
+		"hash", cel2Block.Hash(),
+		"root", cel2Block.Root(),
+		"number", cel2Block.NumberU64(),
+		"gas-used", cel2Block.GasUsed(),
+		"gas-limit", cel2Block.GasLimit(),
+	)
+
+	// If we're not actually writing this to disk, then we're done.
+	if !commit {
+		log.Info("Dry run complete")
+		return nil, nil
+	}
+
+	// Otherwise we need to write the changes to disk. First we commit the state changes.
+	log.Info("Committing trie DB")
+	if err := db.Database().TrieDB().Commit(newRoot, true); err != nil {
+		return nil, err
+	}
+
+	// Next we write the Cel2 genesis block to the database.
+	rawdb.WriteTd(ldb, cel2Block.Hash(), cel2Block.NumberU64(), cel2Block.Difficulty())
+	rawdb.WriteBlock(ldb, cel2Block)
+	rawdb.WriteReceipts(ldb, cel2Block.Hash(), cel2Block.NumberU64(), nil)
+	rawdb.WriteCanonicalHash(ldb, cel2Block.Hash(), cel2Block.NumberU64())
+	rawdb.WriteHeadBlockHash(ldb, cel2Block.Hash())
+	rawdb.WriteHeadFastBlockHash(ldb, cel2Block.Hash())
+	rawdb.WriteHeadHeaderHash(ldb, cel2Block.Hash())
+
+	// TODO(pl): What does finalized mean here?
```

Comment:
I believe it means that there is no chance for the block to be re-orged, which in eth2 means it's got 64 slots or whereabouts that build off it.

Reply:
ANy idea what that means for us then? Is there the same concept of finalization with `op-node`/`op-geth`?

Reply:
For us starting out with a centralised sequencer assuming we wait for deposit transactions on the L1 to become finalized (i.e. 64 blocks deep) then the blocks put out by the sequencer can be considered immediately final (assuming you can execute the block and see it is correct). However I'm not sure how the fraud proving factors into that. BTW this is the [PR](https://github.com/ethereum/go-ethereum/pull/24282) where that method was added.